Do docs ga (#2)
greg-mckeon/cloudflare-docs: products/workers/src/content/runtime-apis/durable-objects.md

# Durable Objects

> Durable Objects are available to anyone with a Workers paid subscription. You can enable them for your account in [the Cloudflare dashboard](https://dash.cloudflare.com/) by navigating to “Workers” and opting-in to pricing.

Reviewer: ```suggestion
Durable Objects are available to anyone with a Workers paid subscription. You can enable them for your account in [the Cloudflare dashboard](https://dash.cloudflare.com/) by navigating to “Workers” and opting in to pricing.
```

> Durable Objects are available to anyone with a Workers paid subscription. You can enable them for your account in [the Cloudflare dashboard](https://dash.cloudflare.com/) by navigating to “Workers” and opting in to pricing.

Reviewer 2: ```suggestion
Durable Objects are available to anyone with a Workers paid subscription. You can enable them for your account in [the Cloudflare dashboard](https://dash.cloudflare.com/) by navigating to **Workers** > **Durable Objects** and opting into pricing.
```

## Background

> Durable Objects provide low-latency coordination and consistent storage for the Workers platform.  A given namespace can support essentially unlimited Durable Objects, with each Object having access to a transactionally-consistent, key-value storage API.

Reviewer: ```suggestion
Durable Objects provide low-latency coordination and consistent storage for the Workers platform.  A given namespace can support essentially unlimited Durable Objects, with each Object having access to a transactionally consistent key-value storage API.
```

  }
}
```
### In-memory state

> Durable Objects can be taken out of memory at any time.  To store state your Durable Object might need in the future to continue processing, use the Transactional Storage API.

Reviewer 3: ```suggestion
A Durable Object may be evicted from memory any time, causing a loss of all transient (in-memory) state.  To persistently store state your Durable Object might need in the future, use the Transactional Storage API.
```

> A Durable Object may be evicted from memory any time, causing a loss of all transient (in-memory) state.  To persistently store state your Durable Object might need in the future, use the Transactional Storage API.

Reviewer 2: Can [Transactional Storage API] have a link out to the place in documentation where it shows how to do this?.

Author: It's the next line, that was why I didn't have a link.

> A Durable Object is given 30 seconds of additional CPU time for every request it processes.  In the absence of failures, in-memory state should not be reset after less than 30 seconds of inactivity.

Reviewer: ```suggestion
A Durable Object is given 30 seconds of additional CPU time for every request it processes, including WebSocket messages. In the absence of failures, in-memory state should not be reset after less than 30 seconds of inactivity.
```

The `fetch()` method of a stub has the exact same signature as the [global `fetch`](/runtime-apis/fetch). However, instead of sending an HTTP request to the internet, the request is always sent to the Durable Object to which the stub points.

Any uncaught exceptions thrown by the Durable Object's `fetch()` handler are propagated to the caller's `fetch()` promise.

> ## Listing Durable Objects

Reviewer: This is notably *not* a runtime API (runtime APIs are the APIs that are natively exposed to Workers at runtime), which is what this page is all about. So it's a little weird to document it here instead of on the using-durable-objects page.

If you really want to keep it on this page, we should at least acknowledge that listing objects is only possible via `api.cloudflare.com`, not via a runtime API.

And if I were someone trying to use this section to figure out how to list my objects, I'd be disappointed when I found that hitting that URL directly didn't work if I didn't know which magic headers I have to provide for my request to be accepted. Are the standard API auth headers documented somewhere else that we could link to here?

Reviewer 3: Whenever I've told people about the API I've always just said "see the top of https://api.cloudflare.com/ for Auth info" and (to my surprise?) it hasn't confused anyone yet.

Reviewer 4: I agree this is an odd place to put it and linking to the top of https://api.cloudflare.com/ is enough for auth.

PR for the v4 api docs is at least merged to staging at this point though, so hopefully less of an issue

Author: this isn't a runtime API, but I would expect to see it called out on this page with a link to the actual docs.  I'll update.

The Cloudflare API supports retrieving a list of Durable Objects within a namespace and a list of namespaces associated with an account.

> The API for listing namespaces is available at https://api.cloudflare.com/client/v4/accounts/$YOUR_ACCOUNT_ID/workers/durable_objects/namespaces

Reviewer 2: ```suggestion
The API for listing namespaces is available [here](https://api.cloudflare.com/client/v4/accounts/$YOUR_ACCOUNT_ID/workers/durable_objects/namespaces).
```

> The API for listing objects in a namespace is available at https://api.cloudflare.com/client/v4/accounts/$YOUR_ACCOUNT_ID/workers/durable_objects/namespaces/$NAMESPACE_ID/objects

Reviewer 2: ```suggestion
The API for listing objects in a namespace is available [here](https://api.cloudflare.com/client/v4/accounts/$YOUR_ACCOUNT_ID/workers/durable_objects/namespaces/$NAMESPACE_ID/objects).
```

The Cloudflare API supports retrieving a list of Durable Objects within a namespace and a list of namespaces associated with an account.

> The API for listing namespaces is available [here](https://api.cloudflare.com/client/v4/accounts/$YOUR_ACCOUNT_ID/workers/durable_objects/namespaces).

Reviewer 2: ```suggestion
The API for listing namespaces is available at https://api.cloudflare.com/client/v4/accounts/$YOUR_ACCOUNT_ID/workers/durable_objects/namespaces.
```

> The API for listing objects in a namespace is available [here](https://api.cloudflare.com/client/v4/accounts/$YOUR_ACCOUNT_ID/workers/durable_objects/namespaces/$NAMESPACE_ID/objects).

Reviewer 4: these shouldn't be hyperlinks, they aren't usefully clickable. The URL itself is the information we're trying to convey, and would need to be populated with account ID and namespace ID as appropriate

Reviewer 2: @koeninger thank you for catching that, I just saw the account ID part.

> The API for listing objects in a namespace is available [here](https://api.cloudflare.com/client/v4/accounts/$YOUR_ACCOUNT_ID/workers/durable_objects/namespaces/$NAMESPACE_ID/objects).

Reviewer 2: ```suggestion
The API for listing objects in a namespace is available at https://api.cloudflare.com/client/v4/accounts/$YOUR_ACCOUNT_ID/workers/durable_objects/namespaces/$NAMESPACE_ID/objects.
```
